### PR TITLE
add new plugin procdump for Linux

### PR DESCRIPTION
**Pull request by EscardosS**
Hello, I made a new plugin procdump for Linux.

How it works (briefly):
- Get process info;
- Collect VMA's info from doubly linked list([Kernel](https://elixir.bootlin.com/linux/v6.0.19/source/include/linux/mm_types.h#L411) 6.1-) or of from maple tree ([Kernel](https://elixir.bootlin.com/linux/v6.1/source/include/linux/mm_types.h#L514) 6.1+);
- Read all VMA's and save them to [core](https://www.gabriel.urdhr.fr/2015/05/29/core-file/) file.

Tested on Kernels 5.10.0 and 6.2.0.

P.S. param `--procdump-use-maple-tree` is used to determine collection method.
P.P.S. Maybe some time later i will add readme file with a more detailed explanation.

### CODE REVIEW

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

**Comment by EscardosS** [comment]
To be honest, I tried to do this, but I stucked with zlib functions linking error of first try. But now i figured how to fix it :)

**Comment by tklengyel** [comment]
> Also it can be done like so:
> 
> ```
> if get_option('plugin-procdump-linux')
> 
>     plugin_sources += 'procdump_linux/procdump_linux.cpp'
> 
>     plugin_sources += 'procdump_linux/writer.cpp'
> 
>     deps += dependency('zlib')
> 
>     config_h.set('ENABLE_PLUGIN_PROCDUMP_LINUX', 1)
> 
> endif
> ```
> 
> Which one is better?

Making the dependency plugin-specific is better, if someone is not using this plugin won't need to satisfy an unused dependency.

**Comment by tklengyel** [comment]
I think it would make sense to merge this plugin into the `procdump2`. We have plugins already that split their logic between windows/linux, like `filetracer`. You largely achieve the same but don't have to use different flags/compile options, the plugin would just detect at runtime which OS specific code-path it needs to use.

File: src/plugins/procdump_linux/procdump_linux.cpp [hunk 1]
```diff
@@ -0,0 +1,865 @@
...
+        vma_list = get_vmas_from_maple_tree(drakvuf, vmi, task->process_data);
+    else
+        vma_list = get_vmas_from_list(drakvuf, vmi, task->process_data);
+
+    calculate_offset(vma_list, &task->note_offset);
+
+    if (vma_list.empty())
+    {
+        print_dump_failure(task->process_base, "Failed to read process memory");
+        //task->writer->finish();
```
[the remark this answers is not in the document]
Comment (tklengyel) [reply src/plugins/procdump_linux/procdump_linux.cpp:722]: Commented out code

File: src/plugins/procdump_linux/procdump_linux.cpp [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+    json_object_put(jobj);
+}
+
+void procdump_linux::dump_zero_page(std::shared_ptr<linux_procdump_task_t> task)
+{
+    uint8_t zeros[VMI_PS_4KB] = {};
+    task->writer->append(zeros, VMI_PS_4KB);
+}
+
+void procdump_linux::read_vm(drakvuf_t drakvuf, vmi_instance_t vmi, vm_area_info vm_area, std::shared_ptr<linux_procdump_task_t> task)
```
[the remark this answers is not in the document]
Comment (tklengyel) [reply src/plugins/procdump_linux/procdump_linux.cpp:168]: This parameter of type is 80 bytes - consider passing a const pointer/reference instead.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

**Comment by tklengyel** [comment]
You can make subfolders inside a plugin to organize them that way. Or just prefix each file with `win_/linux_` to make them easier to categorize quickly.

**Comment by EscardosS** [comment]
Should I change file creation mode to pass CodeQL?

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

**Comment by EscardosS** [comment]
@tklengyel Can you provide the build command line used in Jenkins? Because I don't get any errors during local build

**Pull request closed without merging**